compiler/rustc_infer/src/infer/snapshot/mod.rs

```diff
@@ -99,4 +117,150 @@ impl<'tcx> InferCtxt<'tcx> {
     pub fn opaque_types_added_in_snapshot(&self, snapshot: &CombinedSnapshot<'tcx>) -> bool {
         self.inner.borrow().undo_log.opaque_types_in_snapshot(&snapshot.undo_snapshot)
     }
+
+    pub fn variable_lengths(&self) -> VariableLengths {
+        let mut inner = self.inner.borrow_mut();
+        VariableLengths {
+            type_vars: inner.type_variables().num_vars(),
+            const_vars: inner.const_unification_table().len(),
+            int_vars: inner.int_unification_table().len(),
+            float_vars: inner.float_unification_table().len(),
+            region_vars: inner.unwrap_region_constraints().num_region_vars(),
+        }
+    }
+}
+
+pub struct VariableLengths {
+    region_vars: usize,
+    type_vars: usize,
+    int_vars: usize,
+    float_vars: usize,
+    const_vars: usize,
+}
+
+pub trait NoSnapshotLeaks<'tcx> {
+    type StartData;
+    type EndData;
+    fn snapshot_start_data(infcx: &InferCtxt<'tcx>) -> Self::StartData;
+    fn end_of_snapshot(
+        infcx: &InferCtxt<'tcx>,
+        this: Self,
+        start: Self::StartData,
+    ) -> Self::EndData;
+    fn avoid_leaks(infcx: &InferCtxt<'tcx>, data: Self::EndData) -> Self;
+}
+
+pub trait TrivialNoSnapshotLeaks<'tcx> {}
```

Comment:
Could we use a marker trait for this and opt-out types, consts and lifetimes? Or does that lead to too many overlapping impls pains?

Reply:
we would have to define the marker trait in `rustc_infer`. Defining it earlier, e.g. `rustc_middle` or `rustc_type_ir` would result in overlapping impls pains as e.g. impls for `Result<X, X>` would fail as `Result` may impl the trait.

It also means that we rely on the unstable `auto_traits` feature. Given that we want to share the solver with r-a and they want to use stable, I want to avoid this unless it results in a significant advantage